### PR TITLE
Add new RFC for monorepo

### PR DESCRIPTION
**Pull request by dorianj**
This is a continuation / down-scoping of https://github.com/amundsen-io/rfcs/pull/30. From that discussion, it seems that there is clear support for moving to a mono repo. We'd like to figure out the details of that change and get that rolling and defer the other things that were bundled in the prior RFC.

### CODE REVIEW

**Comment by Golodhros** [comment]
Thanks a lot! 
This looks like the perfect step forward!

File: rfcs/031-monorepo.md
Comment (friendtocephalopods) [inline rfcs/031-monorepo.md:18]: Any reason not to add amundsengremlin to the list as well?

Comment (dorianj) [reply rfcs/031-monorepo.md:18]: Is the intention for `amundsengremlin` to get merged into one of the proxies, or is the long-term intention to leave it as a separate package? I'm not super familiar with the code split there

File: rfcs/031-monorepo.md
Comment (mgorsk1) [inline rfcs/031-monorepo.md:20]: Smooth change management, nice !

However, with current structure, when someone is using the submodule approach can freely choose versions (even refs) for each product separately (I can use frontend in different version than metadata and search). This change somehow limits that capability. It's worth to explicitly mention it in drawbacks.

File: rfcs/031-monorepo.md
Comment (mgorsk1) [reply rfcs/031-monorepo.md:18]: I think that's a point that could have more attention in this rfc. We currently have two approaches for proxy development:
1. 'old' proxies like neo4j, elasticsearch or atlas are kept directly in amundsenmetadata and amundsensearch respectively
2. 'new' proxies like rds or gremlin are separate git repositories.

Maybe we should unify the approach ? afaik the idea with having proxies in different repositories made contributing easier - you could have a team that would be maintainers only for proxy repo, not core amundsen repos. @feng-tao would be more knowledgeable here

File: rfcs/031-monorepo.md
Comment (mgorsk1) [inline rfcs/031-monorepo.md:27]: Another approach (or used alonside changes from this RFC to strengthen the dependency management even more) would be introducing the pip constraints files https://pip.pypa.io/en/stable/user_guide/#constraints-files

File: rfcs/031-monorepo.md
Comment (mgorsk1) [inline rfcs/031-monorepo.md:38]: just a question - do we have any hard data (like results of questionnaires) to back the validity of those issues ? or is it more the result of interacting with users and observation ?

File: rfcs/031-monorepo.md
Comment (Golodhros) [reply rfcs/031-monorepo.md:18]: We could make more extensive use of codeowners files for each package, so we keep the ownership clear.

File: rfcs/031-monorepo.md
Comment (dorianj) [reply rfcs/031-monorepo.md:27]: these are new to me! That's super cool, added.

File: rfcs/031-monorepo.md
Comment (dorianj) [reply rfcs/031-monorepo.md:38]: We don't -- we did the call for feedback in December or January from community members but only got a handful of detailed responses. This is from our interactions with many community members (both longer-time installs) as well as people new to the project. Both groups are impacted, but in different ways.

File: rfcs/031-monorepo.md
Comment (dorianj) [reply rfcs/031-monorepo.md:20]: ah, good point, added

File: rfcs/031-monorepo.md
Comment (dorianj) [reply rfcs/031-monorepo.md:18]: Proposed some language to consider the incubating repos. Open to feedback!

My thought is the separate packages should only be used for incubating, unsupported packages. Once it it officially supported by the project, it should require CR and RFCs from approved maintainers. Otherwise, we will have officially-supported parts of the project that aren't necessarily governed by our normal processes (RFC and CR). If the outside contributor is still making meaningful contributions after the package is landed, we should consider suggesting them as a maintainer (in which case @Golodhros's suggestion of stronger codeowners would make sense -- it might make sense to have that new maintainer start as specialized to the specific area they were previously working on, and if they want to contribute outside of that, we add that on later via codeowner additions)

Comment (feng-tao) [reply rfcs/031-monorepo.md:18]: currently neptune and RDS backend is introduced as part of RFC process. Though I think it is easier for maintain with different package. But I wouldn't call them incubating repos. This is same as Atlas which the engineering team built the whole integration with neo4j while atlas is maintained by @bolkedebruin ING team. Both are officially supported backend in Amundsen for long time.

**Comment by feng-tao** [comment]
as long as different package / artifacts are maintained within single monorepo, I am +1 for the change. But it would be good for @allisonsuarez , @dkunitsk , @jinhyukchang @youngyjd backend ICs from Lyft to take a look and see anything related to existing integrations.

File: rfcs/031-monorepo.md
Comment (dorianj) [reply rfcs/031-monorepo.md:18]: Yeah, that makes sense.

> Though I think it is easier for maintain with different package

Can you elaborate on what is made easier here? I'd like to capture the benefits of those packages without the overhead of actually having separate repos, if that makes sense.

Comment (feng-tao) [reply rfcs/031-monorepo.md:18]: The RFC here is to propose mono repo with different packages. For RDS, gremlin based, it seemed to be easier to maintain them separately given the people who developed those are not part of the core committer team if that makes sense.  It will help to iterate those features faster without asking existing committer to signoff.

But once the repos/features are mature, we should bring them into same repo each of which to be a new package.

**Review by Golodhros — APPROVED** [review]

**Pull request merged**
